streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+import org.apache.kafka.streams.integration.utils.EmbeddedKafkaCluster;
+import org.apache.kafka.streams.integration.utils.IntegrationTestUtils;
+import org.apache.kafka.streams.kstream.Consumed;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.QueryableStoreTypes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+import org.apache.kafka.test.IntegrationTest;
+import org.apache.kafka.test.TestUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(IntegrationTest.class)
+public class OptimizedKTableIntegrationTest {
+    private static final int NUM_BROKERS = 1;
+
+    private static final String INPUT_TOPIC_NAME = "input-topic";
+    private static final String TABLE_NAME = "source-table";
+
+    @Rule
+    public final EmbeddedKafkaCluster cluster = new EmbeddedKafkaCluster(NUM_BROKERS);
+
+    private final Map<KafkaStreams, State> clientStates = new HashMap<>();
+    private final Lock clientStatesLock = new ReentrantLock();
+    private final Condition clientStateUpdate = clientStatesLock.newCondition();
+    private final MockTime mockTime = cluster.time;
+
+    @Before
+    public void before() throws InterruptedException {
+        cluster.createTopic(INPUT_TOPIC_NAME, 2, 1);
+    }
+
+    @After
+    public void after() {
+        for (final KafkaStreams client : clientStates.keySet()) {
+            client.close();
+        }
+    }
+
+    @Test
+    public void standbyShouldNotPerformRestoreAtStartup() throws InterruptedException, ExecutionException {
+        final int numMessages = 10;
+        final int key = 1;
+        final Semaphore semaphore = new Semaphore(0);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder
+            .table(INPUT_TOPIC_NAME, Consumed.with(Serdes.Integer(), Serdes.Integer()),
+                Materialized.<Integer, Integer, KeyValueStore<Bytes, byte[]>>as(TABLE_NAME)
+                    .withCachingDisabled())
+            .toStream()
+            .peek((k, v) -> semaphore.release());
+
+        final KafkaStreams client1 = createClient(builder, streamsConfiguration());
+        final KafkaStreams client2 = createClient(builder, streamsConfiguration());
+        final List<KafkaStreams> clients = Arrays.asList(client1, client2);
+
+        final AtomicLong restoreStartOffset = new AtomicLong(-1);
+        clients.forEach(client -> {
+            client.setGlobalStateRestoreListener(createTrackingRestoreListener(restoreStartOffset));
+            client.start();
+        });
+        waitForClientsToEnterRunningState(clients, 60, TimeUnit.SECONDS);
+
+        produceValueRange(key, 0, 10);
+
+        assertThat("all messages in the first batch were processed in a timely manner",
```

Comment:
Should the error message not point out what went wrong, ie, "messages in the first batch were [not] processed in a timely manner" -- same below

Reply:
I can change that. It reads a little funny "assert that messages in the first batch were not processed in a timely manner", but this should follow project convention.

Reply:
Looking at this again, it is going to confuse the heck out of me if I try to reason about these messages done in the negated form. I'm inclined to pull the messages and perhaps just put a comment above them instead.

Reply:
In most cases we don't have any message, so should be fine to remove. I see your point about `assert that bla` -- however, I think if the assertion hits, the error message reads different (ie, with reversed logic) and hence rephrasing would make it easier to read the error message if it fails (please correct me if I am wrong).

Reply:
I figured I would test it for my own knowledge.

Test case:
```
public class QuickTest {

    @Test
    public void testAssertThatTrueIsEqualToFalse() {
        boolean result = false;
        MatcherAssert.assertThat("the result is true", result, is(equalTo(true)));
    }
}
```

Result:
```
java.lang.AssertionError: the result is true
Expected: is <true>
     but: was <false>
Expected :is <true>
Actual   :<false>
```

I think this actually works out. It is saying the assertion did not hold (AssertionError) and then provides the assertion in the positive form. It also spells out the expected and actual values.

That said, I'm not going to go change all of those back now :).